### PR TITLE
ensure k8s-conform service accounts have objectAdmin rights

### PR DESCRIPTION
**Pull request by spiffxp**
This ensures that for each conformance bucket `k8s-conform-foo`, the same write privileges are given to:
- `group:k8s-infra-conform-foo@kubernetes.io`
- `serviceAccount:service-s390x-k8s@k8s-conform.iam.gserviceaccount.com`

The code before this PR looked like it would probably do this for any newly provisioned GCS bucket.  However `ensure-conformance-storage.sh` was written in a way that it wasn't always refreshing resources and IAM bindings.  So I rewrote it to refresh as much as possible.

While I was here I fixed a few other bugs in the course of troubleshooting / verifying this refactor:
- `ensure-conformance-storage.sh` can no longer provision arbitrary buckets if invoked with a typo'ed argument
- `lib_iam.sh` doesn't crash on empty IAM policies
- "repo" seemed like the wrong term for what we're provision conformance buckets for, so it's now "offering"
- `K8S_INFRA_DEBUG=true ./ensure-whatever.sh` will list and leave behind the contents of `$TMPDIR` instead of removing it
- `lib*.sh` is no longer executable (these files should be sourced, not executed)

Fixes https://github.com/kubernetes/k8s.io/issues/1850

### CODE REVIEW

**Comment by spiffxp** [comment]
Ran `K8S_INFRA_ENSURE_ONLY_SERVICES_WILL_FORCE_DISABLE=true ./ensure-conformance-storage.sh` as of https://github.com/kubernetes/k8s.io/pull/1890/commits/5890959f40acf42b3d19b3b2ba86a5592b5cd300

**Comment by spiffxp** [comment]
It looks like this decided to create new keys for two service accounts, because neither of these had corresponding secrets provisioned beforehand
```sh
for email in $(gcloud iam service-accounts list --project=k8s-conform --filter='email ~ "iam.gserviceaccount.com"' --format='value(email)'); do 
  echo "${email}"...; 
  gcloud iam service-accounts keys list --projct=k8s-conform --iam-account=${email} --managed-by=user; 
done
```
```
service-huaweicloud@k8s-conform.iam.gserviceaccount.com...
KEY_ID                                    CREATED_AT            EXPIRES_AT
ce8a54a39915f738a85a28d41d521b998fac5166  2020-02-27T01:30:44Z  9999-12-31T23:59:59Z
d07efa6acb769eac354120680c011168326368c7  2021-04-08T20:43:11Z  9999-12-31T23:59:59Z
service-inspur@k8s-conform.iam.gserviceaccount.com...
KEY_ID                                    CREATED_AT            EXPIRES_AT
6ffe09fb3aa6a13c62d778eb4d3441b1f14059b7  2021-02-23T06:37:03Z  9999-12-31T23:59:59Z
service-s390x-k8s@k8s-conform.iam.gserviceaccount.com...
KEY_ID                                    CREATED_AT            EXPIRES_AT
045c845ce8d8e1a5bc5671fac07c71353987596c  2020-09-23T21:18:10Z  9999-12-31T23:59:59Z
service-provider-openstack@k8s-conform.iam.gserviceaccount.com...
KEY_ID                                    CREATED_AT            EXPIRES_AT
92b2b472b09b1714aa69fb4742b9ec5776ade47d  2021-02-15T15:18:07Z  9999-12-31T23:59:59Z
service-cri-o@k8s-conform.iam.gserviceaccount.com...
KEY_ID                                    CREATED_AT            EXPIRES_AT
6ac4a3cd42ef8b169502c363f2c32485005c2785  2020-02-26T12:52:14Z  9999-12-31T23:59:59Z
16204156ed33ab4865f83a4ba0766de992e90998  2021-04-08T20:32:12Z  9999-12-31T23:59:59Z
```

The two accounts that have new keys were created before 2020-05-28, when https://github.com/kubernetes/k8s.io/pull/873 added the provision-to-secret pattern.  The previous code gated creating new secrets on service account existence, where the code now gates on secret existence.

I'll open a followup issue to get the key users to switch to the new keys, but I don't consider it a big security risk to leave both keys enabled.

**Comment by spiffxp** [comment]
https://github.com/kubernetes/k8s.io/pull/1892 reflects basically all of the changes these scripts caused.  It was too early to capture the latest run I did in https://github.com/kubernetes/k8s.io/pull/1890#issuecomment-816326889.  I'll point those changes out in the next audit PR

**Review by saschagrunert — APPROVED** [review]

**Comment by saschagrunert** [comment]
/hold 
for further reviews s

**Comment by spiffxp** [comment]
/hold cancel
https://github.com/kubernetes/k8s.io/pull/1897#discussion_r610703658 shows services getting removed

/assign @thockin 
if you want to review post-merge

**Comment by thockin** [comment]
Aaron, your bash is so pleasant to read that I'm going to invite you to collab on my bash-only rewrite of Kubernetes.

/lgtm

File: infra/gcp/lib_gsm.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#   $3: The service-account (e.g. "foo@k8s-infra.iam.gserviceaccount.com")
+function ensure_serviceaccount_key_secret() {
+    if [ ! $# -eq 3 -o -z "$1" -o -z "$2" -o -z "$3" ]; then
+        echo "ensure_serviceaccount_key_secret(project, secret, serviceaccountt) requires 3 arguments" >&2
+        return 1
+    fi
+
+    local project="${1}"
+    local secret="${2}"
+    local serviceaccount="${3}"
```
Comment (ameukam) [inline infra/gcp/lib_gsm.sh:72]: @spiffxp you mean `email` ?

File: infra/gcp/lib_gsm.sh [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+# Ensures a secret exists in the given project with the given name. If the
+# secret does not exist, it is pre-populated with a newly created private key
+# for the given service-account
+# Arguments:
+#   $1: The project id hosting the secret (e.g. "k8s-infra-foo")
+#   $2: The secret name (e.g. "my-secret")
+#   $3: The service-account (e.g. "foo@k8s-infra.iam.gserviceaccount.com")
+function ensure_serviceaccount_key_secret() {
+    if [ ! $# -eq 3 -o -z "$1" -o -z "$2" -o -z "$3" ]; then
+        echo "ensure_serviceaccount_key_secret(project, secret, serviceaccountt) requires 3 arguments" >&2
```
Comment (ameukam) [inline infra/gcp/lib_gsm.sh:66]: `email` instead of `serviceaccountt`

**Pull request merged**